services/src/main/java/org/keycloak/services/resources/SessionCodeChecks.java

```diff
@@ -181,35 +179,26 @@ public AuthenticationSessionModel initialVerifyAuthSession() {
 
         }
 
-        // See if we are already authenticated and userSession with same ID exists.
-        UserSessionModel userSession = authSessionManager.getUserSessionFromAuthCookie(realm);
-
-        boolean authenticating = !CookieHelper.getCookieValue(session, AUTH_SESSION_ID).isEmpty();
+        // Otherwise just try to restart from the cookie
+        RootAuthenticationSessionModel existingRootAuthSession = authSessionManager.getCurrentRootAuthenticationSession(realm);
+        response = restartAuthenticationSessionFromCookie(existingRootAuthSession);
 
-        if (authenticating) {
-            // if there is an auth session, make sure the user is not yet authenticated
+        // if restart from cookie was not found check if the user is already authenticated
+        if (response.getStatus() != Response.Status.FOUND.getStatusCode()) {
             AuthenticationManager.AuthResult authResult = lockUserSessionsForModification(session, () -> authenticateIdentityCookie(session, realm, false));
 
-            if (authResult != null) {
-                userSession = authResult.getSession();
-            }
-        }
+            if (authResult != null && authResult.getSession() != null) {
+                LoginFormsProvider loginForm = session.getProvider(LoginFormsProvider.class).setAuthenticationSession(authSession)
+                        .setSuccess(Messages.ALREADY_LOGGED_IN);
 
-        if (userSession != null) {
-            LoginFormsProvider loginForm = session.getProvider(LoginFormsProvider.class).setAuthenticationSession(authSession)
-                    .setSuccess(Messages.ALREADY_LOGGED_IN);
+                if (client == null) {
+                    loginForm.setAttribute(Constants.SKIP_LINK, true);
```

Comment:
I guess this is not strictly related but a nice improvement to how the page is rendered, right?

Reply:
@pedroigor I have just moved this part from above, so I'm Doing the same than before. :smile:

Reply:
Right, missed it.